### DISCUSSION TITLE
Developed a model from scratch and compared to this

### DISCUSSION THREAD

**Discussion by Vin1001** [amazon-science/chronos-forecasting#49] [Show and tell]
I have developed my own version of OLS and used it in stock price forecasting: https://github.com/Vin1001/TSA

Here's a comparison between the two 👇
![image](https://github.com/amazon-science/chronos-forecasting/assets/120188703/744b53bf-5bda-46f1-bccb-5dac72300969)
![image](https://github.com/amazon-science/chronos-forecasting/assets/120188703/2525e2d1-77b8-4b16-a88f-59c856649eef)

**Comment by abdulfatir** [comment]
Hi! Thanks for sharing! Do you have a specific question you'd like to ask or further insights to share? Regarding these examples, could you share a working example (and the time series) of how you generated these forecasts from both models?

**Comment by Vin1001** [comment]
Hey! Here's my whole notebook where i created my model, the plots i had
provided are near the end of the notebook. The algebra behind it all is
documented in my github repo https://github.com/Vin1001/TSA/

**Reply by abdulfatir** [reply]
Did you forget to attach the notebook? I can't find something like that in your repo.

**Comment by Vin1001** [comment]
The notebook that i provided through mail is a newer version of the
notebook that's available on my repo in the notebooks folder. I can update
that with this one if needed.

**Reply by abdulfatir** [reply]
FYI, if you are attaching any file through email, it's not going through. I only see your text.

**Comment by Vin1001** [comment]
Alright I'll update it in my repo. It'll be visible there. It is named
'autoreg'

**Reply by abdulfatir** [reply]
I checked the code briefly and I am not completely sure what's going on in the `LinearRegression` model. It looks like you're setting your last column as the `target` but the last column is actually a lagged value and I believe the `target` should be column `0`. Can you explain your code?

**Comment by Vin1001** [comment]
The last column will the target since it contains the future value, which
we need to predict. Kindly see the attached note for clarification.

**Reply by abdulfatir** [reply]
As I mentioned before, your attachments are not showing up. Please post via Github web.

**Comment by abdulfatir** [comment]
There were some issues with the inference parameters for Chronos in your code. When in doubt, please use the defaults. I played around a bit and here's what I found.

### Chronos (defaults)
```py
import torch
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from chronos import ChronosPipeline


def getData(name, per, intrvl, cols=["Close"]):
    df = yf.Ticker(name)
    df = pd.DataFrame(df.history(period=per, interval=intrvl))

    return df[cols].copy()


df = getData(name="adanient.ns", per="20d", intrvl="15m")
series = df.reset_index()["Close"]
prediction_length = 32
context, actuals = series[:-prediction_length], series[-prediction_length:]


pipeline = ChronosPipeline.from_pretrained(
    "amazon/chronos-t5-base", device_map="cuda:2"
)
forecast = pipeline.predict(torch.tensor(context), prediction_length=prediction_length)[
    0
].numpy()
forecast = forecast
forecast_index = range(len(context), len(context) + prediction_length)
low, median, high = np.quantile(forecast, [0.25, 0.5, 0.75], axis=0)

plt.figure(figsize=(8, 4))
plt.plot(context, color="royalblue", label="historical data")
plt.plot(forecast_index, actuals, color="forestgreen", label="actuals")
plt.plot(forecast_index, median, color="tomato", label="median forecast")
plt.fill_between(
    forecast_index,
    low,
    high,
    color="tomato",
    alpha=0.3,
    label="50% prediction interval",
)
plt.legend(loc="upper left")
plt.title("Chronos")
plt.grid()
plt.show()
```

#### Output

![image](https://github.com/amazon-science/chronos-forecasting/assets/4028948/e82e13fb-51b6-40a2-8f03-4596e278c9eb)

You might see some quantization artifacts in the forecast above. This happens because the absolute scale of the time series is significantly larger than its variation. One quick fix for the quantization artifacts is to use another type of normalization. Let's try standardization. 

### Chronos (with standardization) 

```py
import torch
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from chronos import ChronosPipeline


def getData(name, per, intrvl, cols=["Close"]):
    df = yf.Ticker(name)
    df = pd.DataFrame(df.history(period=per, interval=intrvl))

    return df[cols].copy()


df = getData(name="adanient.ns", per="20d", intrvl="15m")
series = df.reset_index()["Close"]
prediction_length = 32
context, actuals = series[:-prediction_length], series[-prediction_length:]
mean, scale = context.mean(), context.std()
norm_context = (context - mean) / scale


pipeline = ChronosPipeline.from_pretrained(
    "amazon/chronos-t5-base", device_map="cuda:2"
)
forecast = pipeline.predict(
    torch.tensor(norm_context), prediction_length=prediction_length
)[0].numpy()
forecast = forecast * scale + mean
forecast_index = range(len(context), len(context) + prediction_length)
low, median, high = np.quantile(forecast, [0.25, 0.5, 0.75], axis=0)

plt.figure(figsize=(8, 4))
plt.plot(context, color="royalblue", label="historical data")
plt.plot(forecast_index, actuals, color="forestgreen", label="actuals")
plt.plot(forecast_index, median, color="tomato", label="median forecast")
plt.fill_between(
    forecast_index,
    low,
    high,
    color="tomato",
    alpha=0.3,
    label="50% prediction interval",
)
plt.legend(loc="upper left")
plt.title("Chronos")
plt.grid()
plt.show()
```

#### Output

![image](https://github.com/amazon-science/chronos-forecasting/assets/4028948/e57af27d-558c-4f63-99aa-369418a1b315)


Nice! The quantization artifacts are gone. The forecasts are a bit off but do note that we're unrolling into the entire prediction horizon without feeding in the actual lags as in the case of the linear model. Let's try rolling 1-step predictions.

### Chronos (rolling 1-step)

```py
import torch
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from chronos import ChronosPipeline


def getData(name, per, intrvl, cols=["Close"]):
    df = yf.Ticker(name)
    df = pd.DataFrame(df.history(period=per, interval=intrvl))

    return df[cols].copy()


df = getData(name="adanient.ns", per="20d", intrvl="15m")
series = df.reset_index()["Close"]
prediction_length = 32

pipeline = ChronosPipeline.from_pretrained(
    "amazon/chronos-t5-base", device_map="cuda:2"
)

# rolling prediction
forecast = []
for j in range(prediction_length):
    context = series[: -prediction_length + j]
    mean, scale = context.mean(), context.std()
    norm_context = (context - mean) / scale
    forecast_j = pipeline.predict(
        torch.tensor(norm_context),
        prediction_length=1
    )[0].numpy()
    forecast_j = forecast_j * scale + mean
    forecast.append(forecast_j)
forecast = np.concatenate(forecast, axis=-1)

# for plotting
context, actuals = series[:-prediction_length], series[-prediction_length:]


forecast_index = range(len(context), len(context) + prediction_length)
low, median, high = np.quantile(forecast, [0.25, 0.5, 0.75], axis=0)

plt.figure(figsize=(8, 4))
plt.plot(context, color="royalblue", label="historical data")
plt.plot(forecast_index, actuals, color="forestgreen", label="actuals")
plt.plot(forecast_index, median, color="tomato", label="rolling median forecast")
plt.fill_between(
    forecast_index,
    low,
    high,
    color="tomato",
    alpha=0.3,
    label="50% prediction interval",
)
plt.legend(loc="upper left")
plt.title("Chronos")
plt.grid()
plt.show()
``` 

![image](https://github.com/amazon-science/chronos-forecasting/assets/4028948/314e0c44-72a5-4751-802e-7632348f0229)

**Comment by Vin1001** [comment]
Okay I have uploaded a file named TS_concept.pdf in the notebooks directory

**Reply by abdulfatir** [reply]
I understand what you're trying to do, but please check if you're actually doing that in your code. 
```py
self.y_train[self.new_mat.iloc[:,target-1:target].columns] = np.array(self.new_mat.iloc[:,target-1:target])
self.X_train = self.new_mat.drop(self.y_train.columns, axis=1)
```

In any case, I have answered the issue about Chronos. Let me know if you any other questions _about Chronos_.

**Comment by Vin1001** [comment]
If I wasn't doing it in the code, why would it magically give the expected
result?

**Reply by abdulfatir** [reply]
It's not clear to me from your code. Anyway, let me know if you have any other questions about Chronos. :)

**Comment by Vin1001** [comment]
Its simple data preprocessing. It's not even that i have used high level ML
libraries like sklearn. They're just pandas and numpy modules.

**Reply by abdulfatir** [reply]
Maybe I misunderstood something. I hope my main response addresses your question on Chronos. I am closing this thread now. Please feel free to open another one, if you have further questions.

**Comment by Vin1001** [comment]
I am still ready to explain the whole code line by line if you want. Let's
setup a meeting.
